Add VectorPostprocessingCapability for functor + mesh division + reduction

GiudGiud opened the pull request:
Based on the divisions PR
This will enable SpatialUO based (generalFieldUOTransfer) transfers. MeshDivisions do not have to cover the entire mesh, and the part outside can be excluded. 

I'm totally biased but I like this parameter-based selection a lot. 
Instead of hard-baking the reduction, the operand type and the domain on every object (like we do rn), this is left to the user. 
So 1 object can do:
3 operand types, 3 reduction types (min & max are together in our code already), 10 division types (really can only count 4 in what we have right now: layered, nearest, cylindrical layered, ...)
so 1 object can do the work of 36 objects (on paper, we dont need all 36 in practice)

GiudGiud committed:
Add a VPP for performing reductions on functors based on a mesh division
refs #25901 #22789

GiudGiud committed:
Add interface so we can start transferring mesh-division based quantities

GiudGiud committed:
Fix cylindrical division range to allow 0 min radius
refs #25901

GiudGiud committed:
Add doc and a few tests for mesh division-based reduction
refs #22789 #25901

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1889280/) on 1a359cf wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26099/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1889321/) on 1a359cf wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>268579</th>
      <th colspan=4>#26099 1a359c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.75%</td>
      <td>85.72%</td>
      <td>-0.03%</td>
      <td>80.83%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>93097</td>
      <td>93577</td>
      <td>+480</td>
      <td>485</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15471</td>
      <td>15585</td>
      <td>+114</td>
      <td>115</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26099/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26099/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26099/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 80.83% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented on framework/doc/content/source/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.md:
```diff
@@ -0,0 +1,36 @@
+the returned value of the reduction will be:
+
+- 0 for an average and an integral reduction
+- the maximum floating point number (near `1e300`) for a minimum
+- the minimum floating point number (near `-1e300`) for a maximum
```
should specify what floating point type you're referring to

lindsayad replied:
You may alternatively just not specify the number since `Real` can change depending on the libMesh configuration

lindsayad commented on framework/doc/content/source/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.md:
```diff
@@ -0,0 +1,36 @@
+
+If there are elements within the subdomains of the `MeshDivisionFunctorReductionVectorPostprocessor` that
+are outside the user-specified [!param](/VectorPostprocessors/MeshDivisionFunctorReductionVectorPostprocessor/mesh_division),
+a warning will be output and the element will be ignored.
```
```suggestion
a warning will be output and the elements will be ignored.
```

lindsayad commented on framework/include/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.h:
```diff
@@ -0,0 +1,58 @@
+ * This MeshDivisionFunctorReductionVectorPostprocessor serves to integrate functors based
+ * on the index of the elements in the mesh division
+ */
+class MeshDivisionFunctorReductionVectorPostprocessor : public ElementVectorPostprocessor,
```
Aren't we supposed to use reporters instead of vector postprocessors?

GiudGiud replied:
I dont know? I would think VPPs are fine since:

We can get vectors from VPP using the reporter interface or the VPPInterface
and VPPs are a good concept user-wise

lindsayad replied:
@loganharbour you are a codeowner for reporters. What say you?

lindsayad commented on framework/include/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.h:
```diff
@@ -0,0 +1,58 @@
+  virtual void execute() override;
+  virtual void finalize() override;
+  virtual void threadJoin(const UserObject & uo) override;
+  /// Return Integral values
```
```suggestion

  /// Return Integral values
```

lindsayad commented on framework/include/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.h:
```diff
@@ -0,0 +1,58 @@
+  {
+    return _functor_reductions;
+  };
+  virtual Real spatialValue(const Point & p) const override;
```
```suggestion

  virtual Real spatialValue(const Point & p) const override;
```

lindsayad commented on framework/include/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.h:
```diff
@@ -0,0 +1,58 @@
+  const unsigned int _nfunctors;
+  /// Mesh division providing the division
+  const MeshDivision & _mesh_division;
+  /// Functors that undergo the reduction
```
```suggestion
  /// Functors that are evaluated to create the reduction
```

lindsayad commented on framework/src/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.C:
```diff
@@ -0,0 +1,170 @@
+                                                         "Functors to apply the reduction on");
+  params.addRequiredParam<MeshDivisionName>(
+      "mesh_division",
+      "Mesh division object forming the partition of the elements to perform the reduction within");
```
```suggestion
      "Mesh division object which dictates the elements to perform the reduction with");
```

lindsayad commented on framework/src/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.C:
```diff
@@ -0,0 +1,170 @@
+    _mesh_division(_fe_problem.getMeshDivision(getParam<MeshDivisionName>("mesh_division"), _tid))
+{
+  // Gather the functors
+  const auto functor_names = getParam<std::vector<MooseFunctorName>>("functors");
```
```suggestion
  const auto & functor_names = getParam<std::vector<MooseFunctorName>>("functors");
```

lindsayad commented on framework/src/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.C:
```diff
@@ -0,0 +1,170 @@
+{
+  // Gather the functors
+  const auto functor_names = getParam<std::vector<MooseFunctorName>>("functors");
+  for (const auto i : make_range(_nfunctors))
```
```suggestion
  for (const auto & functor_name : functor_names)
```

lindsayad commented on framework/src/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.C:
```diff
@@ -0,0 +1,170 @@
+  // Gather the functors
+  const auto functor_names = getParam<std::vector<MooseFunctorName>>("functors");
+  for (const auto i : make_range(_nfunctors))
+    _functors.push_back(&getFunctor<Real>(functor_names[i]));
```
```suggestion
    _functors.push_back(&getFunctor<Real>(functor_name));
```

lindsayad commented on framework/src/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.C:
```diff
@@ -0,0 +1,170 @@
+  // Set up reduction vectors
+  const auto ndiv = _mesh_division.getNumDivisions();
+  _volumes.resize(ndiv);
+  for (const auto i : index_range(functor_names))
+  {
+    auto & p = declareVector(functor_names[i]);
```
```suggestion
  for (const auto & functor_name : functor_names)
  {
    auto & p = declareVector(functor_name);
```

lindsayad commented on framework/src/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.C:
```diff
@@ -0,0 +1,170 @@
+    else if (_reduction == ReductionEnum::MAX)
+      std::fill(reduction->begin(), reduction->end(), std::numeric_limits<Real>::min());
+    else
+      mooseAssert(true, "Unknown reduction type");
```
```suggestion
      mooseAssert(false, "Unknown reduction type");
```

lindsayad commented on framework/src/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.C:
```diff
@@ -0,0 +1,170 @@
+  const auto state_arg = determineState();
+  if (hasBlocks(_current_elem->subdomain_id()))
+  {
+    auto index = _mesh_division.divisionIndex(*_current_elem);
```
```suggestion
    const auto index = _mesh_division.divisionIndex(*_current_elem);
```

lindsayad commented on framework/src/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.C:
```diff
@@ -0,0 +1,170 @@
+                   Moose::stringify(*_current_elem));
+      return;
+    }
+    for (unsigned int i = 0; i < _nfunctors; ++i)
```
```suggestion
    for (const auto i : make_range(_nfunctors))
```

lindsayad commented on framework/src/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.C:
```diff
@@ -0,0 +1,170 @@
+    }
+    for (unsigned int i = 0; i < _nfunctors; ++i)
+      if (_functors[i]->hasBlocks(_current_elem->subdomain_id()))
+        for (unsigned int qp = 0; qp < _qrule->n_points(); qp++)
```
```suggestion
        for (const auto qp : make_range(_qrule->n_points())
```

lindsayad commented on framework/src/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.C:
```diff
@@ -0,0 +1,170 @@
+  const MeshDivisionFunctorReductionVectorPostprocessor & sibling =
+      static_cast<const MeshDivisionFunctorReductionVectorPostprocessor &>(s);
+
+  for (unsigned int i_f = 0; i_f < _nfunctors; ++i_f)
```
```suggestion
  for (const auto i_f : make_range(_nfunctors))
```

lindsayad commented on test/tests/vectorpostprocessors/mesh_division_functor_reduction/reduction.i:
```diff
@@ -0,0 +1,108 @@
+    variable = u
+    function = u_fn
+  []
+  [u_fvic]
+    type = FunctionIC
+    variable = u_fv
+    function = u_fn
+  []
```
I thought this wasn't supposed to work anymore after @grmnptr's changes but clearly it does

grmnptr replied:
Yeah for @GiudGiud 's request this is still not erroring. It will error for the linear fv variable.

GiudGiud replied:
If it comes down to it, we could ban it. But I would prefer it kept working if we can

GiudGiud committed:
Address Alex review

- more range-based loops
- more auto in current range-based loops
- remove threading failure from printing in threaded loop

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

Update framework/src/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.C

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1895721/) on 47d71c2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26099/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26099/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2a397e13fdb83e5bf9f5a463f4196eacca670e42`

GiudGiud commented:
umm weird threading failure. trying something

GiudGiud commented on framework/src/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.C:
```diff
@@ -0,0 +1,170 @@
+void
+MeshDivisionFunctorReductionVectorPostprocessor::threadJoin(const UserObject & s)
+{
+  const MeshDivisionFunctorReductionVectorPostprocessor & sibling =
```
so I cant say I understand why but this change causes a threaded failure.
I probably should just print the cpp type of sibling after this line to check?

lindsayad replied:
lol that's big-time crazy town. You're static casting to this type ... how could it be something other than `MeshDivisionFunctorReductionVectorPostprocessor` 😱

GiudGiud replied:
ah yeah data race makes more sense.

why did the threading failure not appear for several runs then appear with this (on several recipes!) then disappear again  though. Threads are so annoying

lindsayad replied:
yea they can't be relied on to be consistently reproducible :-/

lindsayad replied:
so I think we can apply this change again

GiudGiud commented on framework/doc/content/source/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.md:
```diff
@@ -0,0 +1,36 @@
+the returned value of the reduction will be:
+
+- 0 for an average and an integral reduction
+- the maximum floating point number (near `1e300`) for a minimum
```
```suggestion
- the maximum floating point number (a very large positive number, depending on the `Real` floating point number type used) for a minimum
```

GiudGiud commented on framework/doc/content/source/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.md:
```diff
@@ -0,0 +1,36 @@
+
+- 0 for an average and an integral reduction
+- the maximum floating point number (near `1e300`) for a minimum
+- the minimum floating point number (near `-1e300`) for a maximum
```
```suggestion
- the minimum floating point number (minus the number mentioned above) for a maximum
```

GiudGiud commented on framework/doc/content/source/vectorpostprocessors/MeshDivisionFunctorReductionVectorPostprocessor.md:
```diff
@@ -0,0 +1,36 @@
+the returned value of the reduction will be:
+
+- 0 for an average and an integral reduction
+- the maximum floating point number (a very large positive number, depending on the `Real` floating point number type used) for a minimum
```
this probably means the test suite would fail with a different number used :S

lindsayad reviewed (changes requested):
need to fix threading data race

GiudGiud commented:
basically stringify on an elem is not thread safe

GiudGiud commented:
It s this one that s not thread safe
https://mooseframework.inl.gov/docs/doxygen/libmesh/namespacelibMesh_1_1Utility.html#aa99a42a5efab878469e7f58e5c033335

I changed the call. This might come back to haunt us though, want an issue on libmesh?

lindsayad commented:
> This might come back to haunt us though, want an issue on libmesh?

Yea sounds good. We can discuss more over there

GiudGiud commented:
Opened https://github.com/libMesh/libmesh/issues/3725

lindsayad reviewed (approved):
probably will fail clang formatting but with the handy persistent approve I don't have to worry 😄

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1897665/) on 8bc3522 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26099/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26099/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ec85e86826fda2e21ce63bb3a8594ec63bdb07d7`

GiudGiud commented:
beautiful process
